[dtensor][1/N] add torchrec even row-wise sharding example

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120265
* #121217
* #121216
* __->__ #120260
* #121179

**Summary**
our goal is to demonstrate that DTensor's capability to represent TorchRec's parameter sharding. Currently this is done with `ShardedTensor` and theoretically `DTensor` can replace it with minor change.

This PR serves as a start of this effort by adding an example test that represents TorchRec's `ShardingType.ROW_WISE` using DTensor. Note that this PR only covers the even sharding case.

**Test Run**
`torchrun --standalone --nnodes=1 --nproc-per-node=4 torch/distributed/_tensor/examples/torchrec_sharding_example.py -e row-wise`

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

XilunWu committed:
[dtensor] add torchrec row-wise sharding example

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor][1/N] add torchrec row-wise sharding example"


**Summary**
our goal is to demonstrate that DTensor's capability to represent TorchRec's parameter sharding. Currently this is done with `ShardedTensor` and theoretically `DTensor` can replace it with minor change.

This PR serves as a start of this effort by adding an example test that represents TorchRec's `ShardingType.ROW_WISE` using DTensor.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wanchaol reviewed (changes requested):
These are not like tests but rather examples, please put them under `torch/distributed/_tensor/examples/torchrec_sharding_example.py`

XilunWu committed:
Update on "[dtensor][1/N] add torchrec row-wise sharding example"


**Summary**
our goal is to demonstrate that DTensor's capability to represent TorchRec's parameter sharding. Currently this is done with `ShardedTensor` and theoretically `DTensor` can replace it with minor change.

This PR serves as a start of this effort by adding an example test that represents TorchRec's `ShardingType.ROW_WISE` using DTensor.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wanchaol commented on test/distributed/_tensor/test_dtensor_sharding.py:
```diff
@@ -0,0 +1,65 @@
+        global_tensor = torch.randn(*global_tensor_shape)
+
+        # example 1: embedding sharding
+        dtensor = distribute_tensor(
```
DTensor.from_local

wanchaol commented on test/distributed/_tensor/test_dtensor_sharding.py:
```diff
@@ -0,0 +1,65 @@
+        src_dtensor = distribute_tensor(
+            global_tensor, device_mesh, row_wise_sharding_placements
+        )
+        splice_result = row_wise_sharding_placements[0]._split_tensor(
```
no private API should be exposed

XilunWu commented:
we want to show:
1) DTensor sharding in load state dict pre hook
2) DTensor copy in load state dict
3) DTensor construction in _initialize_torch_state()

XilunWu committed:
Update on "[dtensor][1/N] add torchrec row-wise sharding example"


**Summary**
our goal is to demonstrate that DTensor's capability to represent TorchRec's parameter sharding. Currently this is done with `ShardedTensor` and theoretically `DTensor` can replace it with minor change.

This PR serves as a start of this effort by adding an example test that represents TorchRec's `ShardingType.ROW_WISE` using DTensor.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor][1/N] add torchrec row-wise sharding example"


**Summary**
our goal is to demonstrate that DTensor's capability to represent TorchRec's parameter sharding. Currently this is done with `ShardedTensor` and theoretically `DTensor` can replace it with minor change.

This PR serves as a start of this effort by adding an example test that represents TorchRec's `ShardingType.ROW_WISE` using DTensor.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/examples/torchrec_sharding_example.py:
```diff
@@ -0,0 +1,173 @@
+    example_func = name_to_example_code[example_name]
+
+    # set up world pg
+    os.environ["MASTER_ADDR"] = "localhost"
```
if we use torchrun directly we don't need to setup those environments at all (from L138 - L146)

wanchaol commented on torch/distributed/_tensor/examples/torchrec_sharding_example.py:
```diff
@@ -0,0 +1,173 @@
+    #   shard.
+
+    # store each rank of the global ProcessGroup in a tensor
+    global_ranks = torch.arange(world_size)
```
I don't see we are using `global_ranks` anywhere?

wanchaol commented on torch/distributed/_tensor/examples/torchrec_sharding_example.py:
```diff
@@ -0,0 +1,173 @@
+    # display the DTensor's sharding
+    if rank == 0:  # the print result is identical on all ranks
+        print("DTensor's sharding be-like in row-wise sharding:")
+        visualize_sharding(dtensor)
```
I think we can enhance visualize_sharding to only do printing on rank 0 by default, instead of doing it on user code.

wanchaol commented on torch/distributed/_tensor/examples/torchrec_sharding_example.py:
```diff
@@ -0,0 +1,173 @@
+        :,
+    ]
+    # the local shard obtained from both approaches should be identical
+    assert torch.equal(local_shard_spliced, local_shard_spliced)
```
hmmm here you are comparing the same thing with itself?

wanchaol reviewed (commented):
Looks much better! please see inlined comments

XilunWu committed:
Update on "[dtensor][1/N] add torchrec row-wise sharding example"


**Summary**
our goal is to demonstrate that DTensor's capability to represent TorchRec's parameter sharding. Currently this is done with `ShardedTensor` and theoretically `DTensor` can replace it with minor change.

This PR serves as a start of this effort by adding an example test that represents TorchRec's `ShardingType.ROW_WISE` using DTensor.

**Test Run**
`python torch/distributed/_tensor/examples/torchrec_sharding_example.py -e row-wise`

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/examples/torchrec_sharding_example.py:
```diff
@@ -0,0 +1,155 @@
+
+    # display the DTensor's sharding
+    visualize_sharding(
+        dtensor, prompt="DTensor's sharding be-like in row-wise sharding:"
```
```suggestion
        dtensor, prompt="Row-wise sharding example in DTensor"
```

wanchaol reviewed (approved):
lgtm

XilunWu committed:
Update on "[dtensor][1/N] add torchrec even row-wise sharding example"


**Summary**
our goal is to demonstrate that DTensor's capability to represent TorchRec's parameter sharding. Currently this is done with `ShardedTensor` and theoretically `DTensor` can replace it with minor change.

This PR serves as a start of this effort by adding an example test that represents TorchRec's `ShardingType.ROW_WISE` using DTensor. Note that this PR only covers the even sharding case.

**Test Run**
`torchrun --standalone --nnodes=1 --nproc-per-node=4 torch/distributed/_tensor/examples/torchrec_sharding_example.py -e row-wise`

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor][1/N] add torchrec even row-wise sharding example"


**Summary**
our goal is to demonstrate that DTensor's capability to represent TorchRec's parameter sharding. Currently this is done with `ShardedTensor` and theoretically `DTensor` can replace it with minor change.

This PR serves as a start of this effort by adding an example test that represents TorchRec's `ShardingType.ROW_WISE` using DTensor. Note that this PR only covers the even sharding case.

**Test Run**
`torchrun --standalone --nnodes=1 --nproc-per-node=4 torch/distributed/_tensor/examples/torchrec_sharding_example.py -e row-wise`

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]
